### PR TITLE
feat: meancount aggregation for WindowAggregate pushdown in enterprise

### PR DESCRIPTION
**Pull request by lesam**
Small refactors to allow code sharing with InfluxDB Enterprise

- [x] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [x] Rebased/mergeable
- [x] Tests pass

### CODE REVIEW

File: storage/reads/array_cursor_gen_test.go [hunk 1]
```diff
@@ -146,7 +146,7 @@ func TestNewWindowAggregateArrayCursorMonths_Float(t *testing.T) {
 			Type: datatypes.AggregateTypeCount,
 		}
 
-		got, _ := newWindowAggregateArrayCursor(context.Background(), agg, window, &MockFloatArrayCursor{})
+		got, _ := NewWindowAggregateArrayCursor(context.Background(), agg, window, &MockFloatArrayCursor{})
```
Comment (danxmoran) [inline storage/reads/array_cursor_gen_test.go:149]: This file is generated, the source `tmpl` needs updated too

**Review by davidby-influx — CHANGES_REQUESTED** [review]
Per @danxmoran - need to change template file. otherwise looks good.

File: storage/reads/array_cursor_gen_test.go (hunk 1 above)
Comment (lesam) [reply storage/reads/array_cursor_gen_test.go:149]: Done

**Review by davidby-influx — DISMISSED** [review]
LGTM

File: storage/reads/response_writer.go [hunk 2]
```diff
@@ -229,7 +229,10 @@ func (w *ResponseWriter) streamCursor(cur cursors.Cursor) {
 			w.streamBooleanArraySeries(cur)
 		case cursors.StringArrayCursor:
 			w.streamStringArraySeries(cur)
+		case cursors.TwoFloatArrayCursor:
+			w.streamTwoFloatArraySeries(cur)
 		default:
+			//TODO: array
```
Comment (danxmoran) [inline storage/reads/response_writer.go:235]: Does this TODO still need doing?

Comment (danxmoran) [reply storage/reads/response_writer.go:235]: Same for the others below

File: storage/reads/array_cursor.gen.go.tmpldata [hunk 3]
```diff
@@ -43,6 +43,14 @@
 				"Accumulate":"sum += a.Values[rowIdx]; count++",
```
Comment (danxmoran) [inline storage/reads/array_cursor.gen.go.tmpldata:43]: Is there any reason to keep the plain "Mean" aggregations?

File: tsdb/cursors/arrayvalues.gen.go.tmpl [hunk 4]
```diff
@@ -0,0 +1,259 @@
+package cursors
```
Comment (danxmoran) [inline tsdb/cursors/arrayvalues.gen.go.tmpl:1]: Sanity-checking: I don't see a new / changed `arrayvalues.gen.go` matching this file. Was the template just missing, or do we need to re-generate?

File: storage/reads/array_cursor.gen.go.tmpldata (hunk 3 above)
Comment (lesam) [reply storage/reads/array_cursor.gen.go.tmpldata:43]: I think yes. We currently still use plain `Mean` for OSS, and could potentially optimise enterprise to use the plain `Mean` in the fully replicated case.

File: tsdb/cursors/arrayvalues.gen.go.tmpl (hunk 4 above)
Comment (lesam) [reply tsdb/cursors/arrayvalues.gen.go.tmpl:1]: The template was just missing. I've ran `go generate ./tsdb/cursors` with no changes.

File: storage/reads/datatypes/storage_common.proto [hunk 5]
```diff
@@ -165,6 +167,36 @@ message ReadResponse {
...
+      MultiUnsigned unsigned_v = 3;
+      MultiBoolean boolean_v = 4;
+      MultiString string_v = 5;
+    }
+  }
+
+  message MultiPointsFrame {
+    repeated sfixed64 timestamps = 1;
+    repeated AnyPoints value_arrays = 2 [(gogoproto.nullable) = false];
+  }
```
Comment (lesam) [inline storage/reads/datatypes/storage_common.proto:198]: New MultiPointsFrom cc @jsternberg

Comment (jsternberg) [reply storage/reads/datatypes/storage_common.proto:198]: Naming feedback but I like this:

```
message FloatValues {
     repeated double values = 1;
}
message IntegerValues {
     repeated int64 values = 1;
}
message UnsignedValues {
     repeated uint64 values = 1;
}
message BooleanValues {
     repeated bool values = 1;
}
message StringValues {
     repeated string values = 1;
}
message AnyPoints {
     oneof data {
       FloatValues floats = 1;
       IntegerValues integers = 2;
       UnsignedValues unsigned = 3;
       BooleanValues booleans = 4;
       StringValues strings = 5;
     }
}
```

File: storage/reads/aggregate_resultset.go [hunk 6]
```diff
@@ -26,7 +26,7 @@ type windowAggregateResultSet struct {
 // is using the `last` aggregate type, and if it doesn't have a window. If both
 // conditions are met, it returns false, otherwise, it returns true.
 func IsAscendingWindowAggregate(req *datatypes.ReadWindowAggregateRequest) bool {
-	if len(req.Aggregate) != 1 {
+	if len(req.Aggregate) == 0 {
```
[the remark this answers is not in the document]
Comment (lesam) [reply storage/reads/aggregate_resultset.go:29]: I think you're right - gave myself https://github.com/influxdata/influxdb/issues/21330 to fix this in 2.x as well

File: storage/reads/datatypes/storage_common.proto (hunk 5 above)
Comment (lesam) [reply storage/reads/datatypes/storage_common.proto:198]: Updated to this. Feels a bit stutter-y but full words are better than the `_v` stuff.

**Review by danxmoran — APPROVED** [review]

**Comment by lesam** [comment]
Closes https://github.com/influxdata/influxdb/issues/21299

**Pull request merged**